### PR TITLE
Updates gatekeeper crds for v3.5.2

### LINKED ISSUE #7595: upgrade gatekeeper to 3.5 to support K8s 1.22

### PR DESCRIPTION
**Pull request by imharshita**
Signed-off-by: Harshita <harshita.sharma6174@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubermatic/kubermatic/issues/7595

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Update Gatekeeper version to v3.5.2 with new crds Assign, AssignMetadata, MutatorPodStatus and resources MutatingWebhookConfiguration, PodDisruptionBudget
```

### CODE REVIEW

File: pkg/controller/user-cluster-controller-manager/resources/resources/gatekeeper/webhook.go [hunk 1]
```diff
@@ -19,9 +19,9 @@ package gatekeeper
 import (
 	"k8c.io/kubermatic/v2/pkg/resources"
 	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
```
Comment (lsviben) [inline pkg/controller/user-cluster-controller-manager/resources/resources/gatekeeper/webhook.go]: NIT: move it down to the k8s imports group

File: pkg/resources/resources.go [hunk 2]
```diff
@@ -334,8 +317,30 @@ const (
...
 	KubermaticNamespace = "kubermatic"
+	// GatekeeperControllerDeploymentName is the name of the gatekeeper controller deployment
+	GatekeeperControllerDeploymentName = "gatekeeper-controller-manager"
+	// GatekeeperAuditDeploymentName is the name of the gatekeeper audit deployment
+	GatekeeperAuditDeploymentName = "gatekeeper-audit"
+	// GatekeeperWebhookServiceName is the name of the gatekeeper webhook service
+	GatekeeperWebhookServiceName = "gatekeeper-webhook-service"
+	// GatekeeperWebhookServerCertSecretName is the name of the gatekeeper webhook cert secret name
+	GatekeeperWebhookServerCertSecretName = "gatekeeper-webhook-server-cert"
+	PodDisruptionName                     = "gatekeeper-controller-manager"
```
Comment (lsviben) [inline pkg/resources/resources.go]: NIT: add comment

File: pkg/resources/resources.go [hunk 3]
```diff
@@ -381,6 +386,9 @@ const (
 	GatekeeperConfigCRDName = "configs.config.gatekeeper.sh"
 	// GatekeeperConstraintTemplateCRDName defines the CRD name for gatekeeper constraint template objects
 	GatekeeperConstraintTemplateCRDName = "constrainttemplates.templates.gatekeeper.sh"
+	MutatorPodStatusCRDName             = "mutatorpodstatuses.status.gatekeeper.sh"
+	AssignCRDName                       = "assign.mutations.gatekeeper.sh"
+	AssignMetadataCRDName               = "assignmetadata.mutations.gatekeeper.sh"
```
Comment (lsviben) [inline pkg/resources/resources.go]: NIT: add comments

**Comment by mfranczy** [comment]
@harshita-sharma012 the PR looks good to me, fix some comments mentioned by @lsviben.

However, from the other hand, why do we keep CRDs deployment in the code instead of mirroring upstream manifests or using helm charts for deployment? Every change in the CRD must be tracked in the code and copied exactly as is in the upstream manifests. IMO that's very prone to errors. I am not blocking this PR, I am just expressing my opinion and I am curious about the decision made here.

EDIT: I see that we even don't copy the validation schema, we just use field `preserveUnknownFields` to stop the API server pruning fields. That's wrong.

**Comment by mfranczy** [comment]
> @harshita-sharma012 the PR looks good to me, fix some comments mentioned by @lsviben.

However, from the other hand, why do we keep CRDs deployment in the code instead of mirroring upstream manifests or using helm charts for deployment? Every change in the CRD must be tracked in the code and copied exactly as is in the upstream manifests. IMO that's very prone to errors. I am not blocking this PR, I am just expressing my opinion and I am curious about the decision made here.

EDIT: I see that we even don't copy the validation schema, we just use field preserveUnknownFields to stop the API server pruning fields. That's wrong IMO.

As discussed internally, we will try to solve that for the next release.

File: pkg/controller/user-cluster-controller-manager/resources/resources/gatekeeper/pod_disruption.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
```
Comment (mfranczy) [inline pkg/controller/user-cluster-controller-manager/resources/resources/gatekeeper/pod_disruption.go]: could you change to 2021?

File: pkg/crd/kubermatic/v1/cluster.go [hunk 5]
```diff
@@ -22,6 +22,7 @@ import (
 	"fmt"
 
 	providerconfig "github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+
```
Comment (mfranczy) [inline pkg/crd/kubermatic/v1/cluster.go:25]: not needed new line

Comment (mfranczy) [reply pkg/crd/kubermatic/v1/cluster.go:25]: my bad that's a different repo, forget :)

**Pull request merged**